File: vi/codes/java/chapter_computational_complexity/recursion.java
/**
 * File: recursion.java
 * Created Time: 2023-08-24
 * Author: krahets (krahets@163.com)
 */

package chapter_computational_complexity;

import java.util.Stack;

public class recursion {
    /* Đệ quy */
    static int recur(int n) {
        // Điều kiện dừng
        if (n == 1)
            return 1;
        // Đệ quy: gọi đệ quy
        int res = recur(n - 1);
        // Trả về: trả về kết quả
        return n + res;
    }

    /* Mô phỏng đệ quy bằng vòng lặp */
    static int forLoopRecur(int n) {
        // Sử dụng ngăn xếp tường minh để mô phỏng ngăn xếp gọi hàm của hệ thống
        Stack<Integer> stack = new Stack<>();
        int res = 0;
        // Đệ quy: gọi đệ quy
        for (int i = n; i > 0; i--) {
            // Mô phỏng "đệ quy" bằng thao tác "push"
            stack.push(i);
        }
        // Trả về: trả về kết quả
        while (!stack.isEmpty()) {
            // Mô phỏng "trả về" bằng thao tác "pop"
            res += stack.pop();
        }
        // res = 1+2+3+...+n
        return res;
    }

    /* Đệ quy đuôi */
    static int tailRecur(int n, int res) {
        // Điều kiện dừng
        if (n == 0)
            return res;
        // Gọi đệ quy đuôi
        return tailRecur(n - 1, res + n);
    }

    /* Dãy Fibonacci: đệ quy */
    static int fib(int n) {
        // Điều kiện dừng f(1) = 0, f(2) = 1
        if (n == 1 || n == 2)
            return n - 1;
        // Gọi đệ quy f(n) = f(n-1) + f(n-2)
        int res = fib(n - 1) + fib(n - 2);
        // Trả về kết quả f(n)
        return res;
    }

    /* Mã điều khiển */
    public static void main(String[] args) {
        int n = 5;
        int res;

        res = recur(n);
        System.out.println("\nKết quả tổng của hàm đệ quy res = " + res);

        res = forLoopRecur(n);
        System.out.println("\nKết quả tổng khi dùng vòng lặp để mô phỏng đệ quy res = " + res);

        res = tailRecur(n, 0);
        System.out.println("\nKết quả tổng của hàm đệ quy đuôi res = " + res);

        res = fib(n);
        System.out.println("\nSố hạng thứ " + n + " của dãy Fibonacci là " + res);
    }
}

File: vi/codes/kotlin/chapter_computational_complexity/recursion.kt
/**
 * File: recursion.kt
 * Created Time: 2024-01-25
 * Author: curtishd (1023632660@qq.com)
 */

package chapter_computational_complexity.recursion

import java.util.*

/* Đệ quy */
fun recur(n: Int): Int {
    // Điều kiện dừng
    if (n == 1)
        return 1
    // Đi xuống: gọi đệ quy
    val res = recur(n - 1)
    // Đi lên: trả về kết quả
    return n + res
}

/* Mô phỏng đệ quy bằng vòng lặp */
fun forLoopRecur(n: Int): Int {
    // Dùng một ngăn xếp tường minh để mô phỏng ngăn xếp gọi hàm của hệ thống
    val stack = Stack<Int>()
    var res = 0
    // Đi xuống: gọi đệ quy
    for (i in n downTo 0) {
        // Mô phỏng "đệ quy" bằng "push"
        stack.push(i)
    }
    // Đi lên: trả về kết quả
    while (stack.isNotEmpty()) {
        // Mô phỏng "trả về" bằng "pop"
        res += stack.pop()
    }
    // res = 1+2+3+...+n
    return res
}

/* Đệ quy đuôi */
tailrec fun tailRecur(n: Int, res: Int): Int {
    // Thêm từ khóa tailrec để bật tối ưu hóa đệ quy đuôi
    // Điều kiện dừng
    if (n == 0)
        return res
    // Gọi đệ quy đuôi
    return tailRecur(n - 1, res + n)
}

/* Dãy Fibonacci: đệ quy */
fun fib(n: Int): Int {
    // Điều kiện dừng f(1) = 0, f(2) = 1
    if (n == 1 || n == 2)
        return n - 1
    // Gọi đệ quy f(n) = f(n-1) + f(n-2)
    val res = fib(n - 1) + fib(n - 2)
    // Trả về kết quả f(n)
    return res
}

/* Code chạy chính */
fun main() {
    val n = 5
    var res: Int

    res = recur(n)
    println("\nRecursion sum result res = $res")

    res = forLoopRecur(n)
    println("\nUsing iteration to simulate recursion sum result res = $res")

    res = tailRecur(n, 0)
    println("\nTail recursion sum result res = $res")

    res = fib(n)
    println("\nThe ${n}th Fibonacci number is $res")
}